### PR TITLE
Implemented OpenFOAM polyMesh reader with regression test for Hex, Te…

### PR DESCRIPTION
**Pull request by kitamuraa24**
…ts, Wedge, and Pyr elements.

### CODE REVIEW

File: test/python/framework/mesh/read_openfoam_meshes.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    grid = meshgen.Execute()
+    grid.ExportToPVTU("Pyr")
+
+    # Test Wedge
+    meshgen = FromFileMeshGenerator(
+        filename="openfoam_meshes/Wedge"
+    )
+    grid = meshgen.Execute()
+    grid.ExportToPVTU("Wedge")
+
```
Comment (andrsd) [inline test/python/framework/mesh/read_openfoam_meshes.py:43]: General notes on the test. 
- You have a log statement in the OpenFOAM reader, so you may want to match that rather then the `Global cell count`
- You should probably match both the cell and node count
- It would probably make sense to split this into 4 tests and each one would be matching the cell/node count
- Why is the export to VTU needed?

File: framework/mesh/io/foam_io.cc [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+        return false;
+      }
+      else
+      {
+        face_verts.emplace_back(std::move(fv));
+        return true;
+      }
+    };
+    ReadFoamList(in, nfaces, read_face, fname, "faces");
+    if (face_verts.size() != nfaces)
```
[the remark this answers is not in the document]
Comment (kitamuraa24) [reply framework/mesh/io/foam_io.cc:486]: Yes you are right; this was me being sloppy sorry about that.

**Comment by ragusa** [comment]
redo the original text of the PR. Currently, it reads "…ts, Wedge, and Pyr elements."
something got cropped

File: framework/mesh/io/foam_io.cc [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+    { // // ...
+      s.get();
+      s.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+      continue;
+    }
+    if (c2 == '*')
+    { // /* ... */
+      s.get();
+      int prev = 0, cur = 0;
+      while (s and not(prev == '*' and (cur = s.get()) == '/'))
```
Comment (wdhawkins) [inline framework/mesh/io/foam_io.cc:39]: This looks like it might result in a hang if prev != '*'.

**Comment by kitamuraa24** [comment]
Give me a moment; I found another bug that I am currently working on fixing that was not present earlier.

**Comment by kitamuraa24** [comment]
Fixed the issue. Redid the ReadCellZones function

**Comment by kitamuraa24** [comment]
Hi all, just wondering if everything is good to merge?

**Comment by wdhawkins** [comment]
Can you rebase this on main and extend one of your tests to run a very simple problem? Just to make sure that boundary conditions, etc. work?
